### ISSUE TITLE
[Sema] Make a vector mask structural: blend vectors of matching lane count and element width

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1083]
Baselines: Align `8c8bfbc7a3169e84ecc8415f5149ab8c61afe863` (alignc 0.7.5, LLVM 22.1.8), Apple M1,
`--profile release`, default `--target-cpu baseline`. Language question first; align-llm figures are
evidence only.

## Problem

`Ty::Mask` carries the *element type* of the vector that produced it (`Mask(Scalar, u32)`,
`crates/align_sema/src/lib.rs:491`), and `select` requires the mask's element type to be identical to the
blended vectors'. But a mask has no element type at the machine level: every `maskN<T>` is `<N x i1>`, and
a lane-wise blend depends only on the lane count and the lane width.

The result is that the single most useful masked idiom in the language — "compare floats, keep the
*indices* of the winners" — does not type-check. That covers argmax/argmin, top-k, first/last match,
histogram bucketing, index compaction, and any "select the payload for the lanes that won" pattern. For a
data-oriented language that promises ordinary slice code lowers to SIMD, this is a hole in the explicit
SIMD surface, independent of any product.

## Evidence

The exact diagnostics, from the pinned `alignc`:

```align
pub fn m(a: vec4<f32>, b: vec4<f32>, x: vec4<i32>, y: vec4<i32>) -> vec4<i32> {
  mm := a > b
  return select(mm, x, y)
}
```

```
error: 'select' mask mask4<f32> does not match the vectors vec4<i32>
```

Identical with `vec4<u32>`:

```
error: 'select' mask mask4<f32> does not match the vectors vec4<u32>
```

The check is a single `match` arm — `crates/align_sema/src/lib.rs:55835-55841`, inside `check_select`:

```rust
        let mc_ty = self.resolve(mc.ty);
        match mc_ty {
            Ty::Mask(ms, mn) if ms == s && mn == n => {}
            Ty::Mask(..) => {
                self.diags.error(format!("'select' mask {} does not match the vectors {}", ty_name(mc_ty), ty_name(Ty::Vec(s, n))), m.span);
                return err;
            }
```

Codegen is already type-agnostic — it never looks at the mask's element type
(`crates/align_codegen_llvm/src/lib.rs:11489-11496`):

```rust
            Rvalue::Select { cond, a, b } => {
                // A `mask` cond (`<N x i1>`, from `select(mask, a, b)`) blends two vectors lane-wise;
                // a scalar `i1` cond (branchless `where`) blends two scalars.
                if matches!(self.f.operand_ty(cond), Ty::Mask(..)) {
                    let c = self.operand(cond)?.into_vector_value();
                    let av = self.operand(a)?.into_vector_value();
                    let bv = self.operand(b)?.into_vector_value();
                    self.builder.build_select(c, av, bv, "vsel").map_err(|e| self.err(e))?
```

So the restriction is entirely a front-end one, and it is stricter than both the machine and the IR.

### The workaround and its cost

The only way to write a vectorized argmax today is to keep the index vector in `f32` so the mask's element
type matches — exact only for indices below 2^24:

```align
  // Fully vectorized argmax over slice<f32>. Workaround for two gaps:
  //  (a) a mask from an f32 compare cannot select an integer vector, so the index vector is f32
  //      (exact for any vocabulary below 2^24);
  //  (b) a lane index must be a literal, so the horizontal step is unrolled by hand.
  mut vidx: vec4<f32> := [0.0 as f32, 1.0 as f32, 2.0 as f32, 3.0 as f32]
  ...
    m := v > vbest
    vbest = select(m, v, vbest)
    vidx = select(m, vcur, vidx)
```

Compiled with the pinned `alignc` at `--target-cpu baseline`, this does produce good code — the inner loop
is 11 instructions per 4 elements:

```
0000000000000040	add	x1, x0, #0x4
0000000000000044	cmp	x1, x2
0000000000000048	b.gt	0xdc            ; surviving range check
000000000000004c	ldr	q4, [x10], #0x10
0000000000000050	fadd.4s	v3, v3, v2
0000000000000054	cmp	x1, x9
0000000000000058	mov	x0, x1
000000000000005c	fcmgt.4s	v5, v4, v1
0000000000000060	bit.16b	v1, v4, v5
0000000000000064	bit.16b	v0, v3, v5
0000000000000068	b.lo	0x40
```

and measures **47.5 µs/call** against **190.4 µs/call** for the scalar `if xs[i] > best` argmax over the
same already-typed `slice<f32>` (152,064 elements, best of 3, same host) — **4.0×**. So the codegen is
fine; what is wrong is that a correct program (integer indices, exact for all lengths) is rejected and the
only accepted program silently loses exactness above 2^24. A language should not make the *correct*
spelling the one that does not compile.

The second gap visible in the same comment is separate but adjacent:

```
error: a vector lane index must be a constant in 0..4
```

(`crates/align_sema/src/lib.rs:47961`, `:60463`) — which forces the horizontal tail to be hand-unrolled.
It is mentioned here for context, not proposed for change in this issue.

## Root cause

`Ty::Mask(Scalar, u32)` over-specifies. The mask's element type is recorded because the comparison that
produced it had one, but nothing downstream needs it: a mask's only semantic content is its lane count
(and, for a legal blend, the lane *width* of the operands it gates).

## Proposed solution

Make the mask rule structural, so there is one rule rather than an exception list.

**Preferred: relax the compatibility predicate to lane count and lane bit width.** Replace the
`ms == s && mn == n` arm at `crates/align_sema/src/lib.rs:55838` with a predicate on
`(lane_count, lane_bits)`:

- `mask4<f32>` may gate `vec4<i32>`, `vec4<u32>`, `vec4<f32>` (all 32-bit lanes, 4 lanes);
- `mask2<f64>` may gate `vec2<i64>`/`vec2<u64>`/`vec2<f64>`;
- `mask4<f32>` may **not** gate `vec4<i64>` or `vec8<i32>` — lane width and lane count must both agree,
  which is exactly what the target's bit-select instruction requires.

This is a pure widening of what type-checks: every program that compiles today still compiles, with
identical IR. Codegen needs no change (verified above).

**Alternative, if the mask's element type is wanted for diagnostics:** keep `Ty::Mask(Scalar, u32)` and add
an explicit, total conversion `mask.reinterpret<T>() -> maskN<T>` legal exactly when the widths agree. This
is more ceremony for the same guarantee and adds vocabulary rather than removing an exception, so the first
form is better for a language that wants one way to do things — but it is a defensible choice if masks are
meant to stay nominally typed.

Either way, the rule should be stated once and apply uniformly to `select`, to any future masked
load/store, and to masked pipeline stages, rather than being special-cased at `check_select`.

## Acceptance criteria

- `select(a > b, x, y)` type-checks for every combination of `f32`/`i32`/`u32` vectors of the same lane
  count, and likewise for the 64-bit and 16-bit/8-bit lane families; a lane-count or lane-width mismatch
  still produces the existing diagnostic.
- Emitted IR for the relaxed cases is a plain `select <N x i1>` (no conversion instruction introduced).
- A masked argmax with an **integer** index vector lands in `examples/` — exact for all lengths, no 2^24
  restriction — and a codegen owner test asserts its inner loop contains `fcmgt.4s` and two `bit.16b`
  (aarch64) / the equivalent `cmpps` + `blendvps` pair (x86-64).
- A regression test asserts the negative cases still fail: `mask4<f32>` gating `vec2<f64>`, and
  `mask4<f32>` gating `vec8<i32>`.
- `crates/align_driver/tests/vectorize_shapes.rs` and `examples/vec_simd.align` are unaffected.

## Relation to existing issues/plans

- New. Nothing in `docs/` discusses mask typing; `examples/vec_simd.align` covers only same-type
  elementwise operations.
- Relevant to **#1064** (revised to Option A — safe typed `slice<u8>` → `slice<T>` views): once logits are
  a `slice<f32>`, the *explicit* SIMD route to an argmax over them needs exactly this relaxation, and
  #1064 deliberately (and correctly) does not add an `argmax` intrinsic. This issue is what makes the
  hand-written vectorized argmax expressible without the `f32`-index trick.
- Independent of the loop-facts contract issues (#1079, #1080, #1081) and of the
  floating-point reduction semantics issue (#1082); it removes a front-end restriction rather than
  changing any lowering.
- No overlap with #1047, #1063, #1065, #1066.

## Impact framing

Measured 4.0× (190.4 → 47.5 µs/call) on a 152,064-element argmax, which in align-llm today sits in the
Align-owned CPU kernels
(greedy logits selection 242 µs/token over 151,936 logits, sampler 267 µs/token over 152,064, same host) —
and Align-compiled code is ~0.1 % of CPU in
a real Metal generation profile, so this is evidence, not a throughput headline. The language case is the
stronger one: the explicit SIMD surface currently rejects the correct spelling of an entire family of
masked-index algorithms and accepts only an approximate one, and the fix is a one-arm relaxation of a
predicate that the backend already ignores.
